README.md
# SatGPT

GPT for satellite mission planning. In development. Get your API keys ready!

SatGPT heavily leverages Shell GPT, a command-line UI for OpenAI's completion API, 
and its features for custom role creation (as of 04/16/2023). Currently for SatGPT to work,
we need to overwrite the default shell roll using the instructions below.

## Installation

1. Clone this repo. Installing from source will be more straightforward than PyPi install at this stage.

2. From inside repo root, install from source.

```
python -m pip install -e . 
```

3. Overwrite `shell-gpt` default `shell` role with `satgpt` alpha demo.

```
python src/satgpt/make_role.py
```
## Usage

1. Start the REPL in shell mode.

```
sgpt --shel --repl temp
```

2. Request a STAC search using natural language

```
>>> find all landsat 8&9 imagery over washington, DC in April 2023
```

3. Request edits or additional steps

```
>>> pipe the results to stacterm cal
```

4. Try other things!

CHANGELOG.rst
Version 0.0.4:
- Moved role creation to manual script run
- Populated README with basic usage

Version 0.0.3:
- Fixed custom role creation for satgpt to grab target machine os/ shell
- Added pystac as install dependency for demos
- Added this CHANGELOG

Version 0.0.2:
- Integrated shell-gpt as a dependency using the custom role feature released in version 0.9.0

tests/test_module_demo.py
#!/usr/bin/env python
# encoding: utf-8

import unittest

from satgpt.module_demo import print_something
import sgpt #works

class TestModuleDemo(unittest.TestCase):
    def test_print_something(self):
        status = print_something("My value")
        self.assertEqual(status, "My value")


src/satgpt/make_satgpt_role.py
import json
from os import getenv

# read in local json file as string
def read_json_file(file_path):
    with open(file_path, "r") as f:
        return f.read()


STAC_COLLECTION = read_json_file(file_path="reference/stac/collection.json")
STAC_LANDSAT = read_json_file(file_path="reference/stac/landsat.json")
STAC_SENTINEL = read_json_file(file_path="reference/stac/sentinel.json")
RIO_SNIPPET = read_json_file(file_path="reference/rio/snippet.txt")
CLIENT_SNIPPET = read_json_file(file_path="reference/stac/client-snippet.txt")

SHELL_PROMPT = """
____

Here are some particular CLI packages to reccomend and their use cases:

stac-client
Important: when using stac-client DO NOT format bounding boxes like: --bbox 102.5,22.5,103.5,23.5
The correct format is: --bbox 102.5 22.5 103.5 23.5
Important: do not return a stac-client command without providing a URL. If the user does not specify, use https://landsatlook.usgs.gov/stac-server/
The cannonical stac-client command is: stac-client search URL --bbox BBOX --datetime DT --query QUERY
An example is:
stac-client search https://landsatlook.usgs.gov/stac-server/  --bbox 102.5 22.5 103.5 23.5 --collections landsat-c2l2-sr --datetime 2022-09-01/2023-04-30 --query "eo:cloud_cover<20"
IMPORTANT: NO COMMAS IN BETWEEN NUMERIC BBOX VALUES. USE SPACES ONLY BETWEEN BBOX VALUES. NO EXECPTIONS.

____

"""

# stac
# # example stac items
# # the following are examples of json stac items. they do not follow actual catalog names or filenames, these must be queried from the
# # catalogs or items specified by the user, most often using jq. for exampple, to get the band names in an item it's keys must be traversed.
# # the heirarchy represented in the examples below is a good guide for how to traverse the json responses.

# # COLLECTION EXAMPLE
# {STAC_COLLECTION}
# # LANDSAT EXAMPLE
# {STAC_LANDSAT}
# # SENTINEL EXAMPLE
# {STAC_SENTINEL}
# ____
# """.format(
#     STAC_COLLECTION=STAC_COLLECTION,
#     STAC_LANDSAT=STAC_LANDSAT,
#     STAC_SENTINEL=STAC_SENTINEL,
# )


# rio
# # rio is a command line tool for writing, reading, and manipulating geospatial raster data. it is the CLI for rasterio.
# # refer to rasterio docs for more details.
# {RIO_SNIPPET}

# Request: """

file_path = getenv("HOME") + "/.config/shell_gpt/roles/shell.json"

with open(file_path, "r") as f:
    role_dict = json.loads(f.read())

file_path = getenv("HOME") + "/.config/shell_gpt/roles/satgpt.json"

role_dict["role"] += SHELL_PROMPT
with open(getenv("HOME") + "/.config/shell_gpt/roles/satgpt.json", "w") as f:
    f.write(json.dumps(role_dict))


src/satgpt/make_role.py
#!/usr/bin/env python
import json
import os
import platform
from os import getenv, pathsep
from os.path import basename


# adapted from https://github.com/TheR1D/shell_gpt/blob/main/sgpt/role.py
def make_satgpt_role():
    from distro import name as distro_name

    operating_systems = {
        "Linux": "Linux/" + distro_name(pretty=True),
        "Windows": "Windows " + platform.release(),
        "Darwin": "Darwin/MacOS " + platform.mac_ver()[0],
    }
    current_platform = platform.system()

    os_name = operating_systems.get(current_platform, current_platform)
    if current_platform in ("Windows", "nt"):
        is_powershell = len(getenv("PSModulePath", "").split(pathsep)) >= 3
        shell_name = "powershell.exe" if is_powershell else "cmd.exe"
    else:
        shell_name = basename(getenv("SHELL", "/bin/sh"))

    SATGPT_ROLE = """###
    Provide only {shell} commands for {os} without any description.
    If there is a lack of details, provide most logical solution.
    Ensure the output is a valid shell command.
    If multiple steps required try to combine them together.
    ____

    Here are some particular CLI packages to reccomend and their use cases:

    stac-client
    Important: when using stac-client DO NOT format bounding boxes like: --bbox 102.5,22.5,103.5,23.5
    The correct format is: --bbox 102.5 22.5 103.5 23.5
    Important: do not return a stac-client command without providing a URL. If the user does not specify, use https://landsatlook.usgs.gov/stac-server/
    The cannonical stac-client command is: stac-client search URL --bbox BBOX --datetime DT --query QUERY
    An example is:
    stac-client search https://landsatlook.usgs.gov/stac-server/  --bbox 102.5 22.5 103.5 23.5 --collections landsat-c2l2-sr --datetime 2022-09-01/2023-04-30 --query "eo:cloud_cover<20"
    IMPORTANT: NO COMMAS IN BETWEEN NUMERIC BBOX VALUES. USE SPACES ONLY BETWEEN BBOX VALUES. NO EXECPTIONS.

    ____
    """.format(
        shell=shell_name, os=os_name
    )

    satgpt_json = {
        "name": "shell",
        "expecting": "Command",
        "variables": {
            "shell": shell_name,
            "os": os_name,
        },
        "role": SATGPT_ROLE,
    }

    file_path = getenv("HOME") + "/.config/shell_gpt/roles/shell.json"

    # make directory if it doesn't exist
    os.makedirs(os.path.dirname(file_path), exist_ok=True)
    with open(file_path, "w") as f:
        f.write(json.dumps(satgpt_json, indent=4))


if __name__ == "__main__":
    # setup for satgpt
    make_satgpt_role()
